reconcilers/childset.go

```diff
@@ -0,0 +1,359 @@
+	// with actual child resources. Non-deterministic IDs will result in the rapid deletion and
+	// creation of child resources.
+	OurChild func(resource Type, child ChildType) (bool, string)
+
+	// Sanitize is called with an object before logging the value. Any value may
+	// be returned. A meaningful subset of the resource is typically returned,
+	// like the Spec.
+	//
+	// +optional
+	Sanitize func(child ChildType) interface{}
+
+	stamp          *ResourceManager[ChildType]
+	lazyInit       sync.Once
+	voidReconciler *ChildReconciler[Type, ChildType, ChildListType]
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) init() {
+	r.lazyInit.Do(func() {
+		var nilCT CT
+		if internal.IsNil(r.ChildType) {
+			r.ChildType = newEmpty(nilCT).(CT)
+		}
+		if internal.IsNil(r.ChildListType) {
+			var nilCLT CLT
+			r.ChildListType = newEmpty(nilCLT).(CLT)
+		}
+		if r.Name == "" {
+			r.Name = fmt.Sprintf("%sChildSetReconciler", typeName(r.ChildType))
+		}
+		r.stamp = &ResourceManager[CT]{
+			Name:                     r.Name,
+			Type:                     r.ChildType,
+			Finalizer:                r.Finalizer,
+			TrackDesired:             r.SkipOwnerReference,
+			HarmonizeImmutableFields: r.HarmonizeImmutableFields,
+			MergeBeforeUpdate:        r.MergeBeforeUpdate,
+			Sanitize:                 r.Sanitize,
+		}
+		r.voidReconciler = r.childReconcilerFor(nilCT, nil, "")
+	})
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) SetupWithManager(ctx context.Context, mgr ctrl.Manager, bldr *builder.Builder) error {
+	r.init()
+
+	c := RetrieveConfigOrDie(ctx)
+
+	log := logr.FromContextOrDiscard(ctx).
+		WithName(r.Name).
+		WithValues("childType", gvk(r.ChildType, c.Scheme()))
+	ctx = logr.NewContext(ctx, log)
+
+	if err := r.validate(ctx); err != nil {
+		return err
+	}
+
+	if err := r.voidReconciler.SetupWithManager(ctx, mgr, bldr); err != nil {
+		return err
+	}
+
+	if r.Setup == nil {
+		return nil
+	}
+	return r.Setup(ctx, mgr, bldr)
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) childReconcilerFor(desired CT, desiredErr error, id string) *ChildReconciler[T, CT, CLT] {
+	return &ChildReconciler[T, CT, CLT]{
+		Name:               id,
+		ChildType:          r.ChildType,
+		ChildListType:      r.ChildListType,
+		SkipOwnerReference: r.SkipOwnerReference,
+		DesiredChild: func(ctx context.Context, resource T) (CT, error) {
+			return desired, desiredErr
+		},
+		ReflectChildStatusOnParent: func(ctx context.Context, parent T, child CT, err error) {
+			result := retrieveChildSetResult[CT](ctx)
+			result.Children = append(result.Children, ChildSetPartialResult[CT]{
+				Id:    id,
+				Child: child,
+				Err:   err,
+			})
+			stashChildSetResult(ctx, result)
+		},
+		HarmonizeImmutableFields: r.HarmonizeImmutableFields,
+		MergeBeforeUpdate:        r.MergeBeforeUpdate,
+		ListOptions:              r.ListOptions,
+		OurChild: func(resource T, child CT) bool {
+			ours, ourId := r.OurChild(resource, child)
+			return ours && (id == ourId || id == "")
```

Comment:
`id == ""`? I only know `voidReconciler` uses empty string as ID and still not sure why this `|| id == ""` is needed.

Reply:
it's so we can call ourChild on the voidReconciler to pick up the default behavior for ChildReconciler before the custom function is invoked. There are guards in place elsewhere to make sure an empty string is never actually used as an id.